### DISCUSSION TITLE
List of repositories as a yaml file

### DISCUSSION THREAD

**Discussion by simone-dandreta** [renovatebot/renovate#22431] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.0.0

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Hello,
in my config.js file I have:

`repositories: ['list of repositories'],`

which is working fine. However as my team will be administering Renovate and we'll onboard hundreds of repos, we find out that it's not scalable to add repositories in a list like the one above. 
Ideally we would like to add them in a yaml file and then reference that yaml file in the repositories array in the config.js file. 
Is it possible?
Thank you.

### What you tried so far.

I can't use the autodiscover filter as it's quite difficult to extract a common path and build a regex out of it.

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by rarkins** [comment]
It's a js file so you can read in that list anyway you want, including fs.readFileSync

**Reply by simone-dandreta** [reply]
Thank you for the answer @rarkins :) 
 That's what I did:
I added `const repos = require('./repositories.json');
`
at the beginning of the config.js file. The repositories.json file is in the root folder, the same as config.js.
I keep getting:

`DEBUG: Checking for config file in /github-action/config.js
FATAL: No custom config file found on disk`

but both files are definitely present and in the same directory. I am running this as a Github action and using the renovatebot:slim image. I am wondering if this is because the Renovate image doesn't have the repositories.json included? If that's the case, how can I add it to the configuration?

**Reply by rarkins** [reply]
How exactly are you running renovate?

**Reply by rarkins** [reply]
Was it working before you added the require?

**Reply by simone-dandreta** [reply]
I am using the https://github.com/renovatebot/github-action.
It was working fine before I started to use a separate json file. I made a step ahead but now it tells me that:

`FATAL: Fatal error: config.repositories is not iterable
       "err": {
         "message": "config.repositories is not iterable",
         "stack": "TypeError: config.repositories is not iterable\n    at validateConfigSecrets (/opt/buildpack/tools/renovate/35.100.3/node_modules/renovate/lib/config/secrets.ts:46:37)\n    at /opt/buildpack/tools/renovate/35.100.3/node_modules/renovate/lib/workers/global/index.ts:147:28\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at start (/opt/buildpack/tools/renovate/35.100.3/node_modules/renovate/lib/workers/global/index.ts:123:5)\n    at /opt/buildpack/tools/renovate/35.100.3/node_modules/renovate/lib/renovate.ts:18:22"
       }`

My repositories.json file looks like:

`{
	"repositories": [
		"Sentinel-One/gitops-hello-world",
		"Sentinel-One/docker-base-images",
		"Sentinel-One/k8s-deployments-checks",
		"Sentinel-One/fedramp-replication",
		"Sentinel-One/atlas-auth-service"
	]
}  `

and in the config.js I have:
```

const repos = JSON.parse(process.env.RENOVATE_REPOSITORIES || '[]');

console.log('Parsed repos:', repos);
console.log('Parsed repositories:', repos.repositories);

module.exports = {
  repositories: repos.repositories,
```

I guess I am iterating through the object but if i look at the printed output, repos.repositories is:
`[
  'Sentinel-One/gitops-hello-world',
  'Sentinel-One/docker-base-images',
  'Sentinel-One/k8s-deployments-checks',
  'Sentinel-One/fedramp-replication',
  'Sentinel-One/atlas-auth-service'
]`

which seems to me an array. Do you have any idea what's going on please?

**Reply by rarkins** [reply]
Your `repositories.json` probably needs to start and end like `[ ... ]` because it's an array, not an object. I can't tell for sure because you haven't shared your new `config.js`

**Reply by viceice** [reply]
the GitHub action doesn't mount the json file to the container, so the require in config.js fails.

**Reply by simone-dandreta** [reply]
@viceice I mounted the file as environment variable and now that error is gone. 
@rarkins here is my config.js:

```
const repos = JSON.parse(process.env.RENOVATE_REPOSITORIES || '[]');

console.log('Parsed repos:', repos);
console.log('Parsed repositories:', repos.repositories);

module.exports = {
  repositories: repos.repositories,
  onboardingConfig: {"$schema": "https://github.com/Sentinel-One/renovate-configure/blob/master/renovate-schema.json"},
  extends: ["github>Sentinel-One/renovate-configure"],
  username: 'REDACTED',
  gitAuthor:"REDACTED",
  "hostRules": [
    {
      "matchHost": "REDACTED",
      "hostType": "docker",
      "username":  "REDACTED",
      "password":REDACTED,      
    }
  ]
}
```

And I see in the renovate logs that the config at certain point is:

```

DEBUG: Env config
       "config": {
         "hostRules": [],
         "token": "***********",
         "repositories": {
           "repositories": [
             "Sentinel-One/gitops-hello-world",
             "Sentinel-One/docker-base-images",
             "Sentinel-One/k8s-deployments-checks",
             "Sentinel-One/fedramp-replication",
             "Sentinel-One/atlas-auth-service"
           ]
         }
       }
```

which I don't think it's correct. However the console.log shows that the repos.repositories variable is:

```
[
  'Sentinel-One/gitops-hello-world',
  'Sentinel-One/docker-base-images',
  'Sentinel-One/k8s-deployments-checks',
  'Sentinel-One/fedramp-replication',
  'Sentinel-One/atlas-auth-service'
]
```

which seems to me correct. Something is injecting the repository object and I am not sure why.

**Reply by rarkins** [reply]
`RENOVATE_REPOSITORIES` is a reserved keyword. You should avoid using any `RENOVATE_` variables yourself. If you use that variable you don't need to do any parsing of your own, i.e. do this:

```
module.exports = {
  onboardingConfig: {"$schema": "https://github.com/Sentinel-One/renovate-configure/blob/master/renovate-schema.json"},
  extends: ["github>Sentinel-One/renovate-configure"],
  username: 'REDACTED',
  gitAuthor:"REDACTED",
  "hostRules": [
    {
      "matchHost": "REDACTED",
      "hostType": "docker",
      "username":  "REDACTED",
      "password":REDACTED,      
    }
  ]
}
```

**Reply by simone-dandreta** [reply]
I am confused. Should I use the RENOVATE_REPOSITORIES or not? And looking at the config.js you posted, how does that variable gets populated? How does it know it has to look for a repositories.json file?

**Reply by simone-dandreta** [reply]
Ah I see what you mean. Looking at the doc, it looks like the RENOVATE_REPOSITORIES is a list of comma separated variables. So if I use it, I think I don't have to use any json file but just a plain text file with a list of repos. Or I can try to inject a json?

**Reply by simone-dandreta** [reply, marked as answer]
OK looks like I made it working by using a REPOSITORIES variable. Thanks everybody for your help!
